File: MADDriver.source/MADDriver.Source/MyDebugStr.c
/*
 *  MADDebugStr.c
 *  PPMacho
 *
 *  Created by C.W. Betts on 11/18/09.
 *  Copyright 2009 __MyCompanyName__. All rights reserved.
 *
 */

#include "MAD.h"
#include <stdio.h>

static void(*privateDebugStr)(short, const char*, const char*) = NULL;

extern void MADRegisterDebugFunc(void(*adebug)(short, const char*, const char*))
{
	privateDebugStr = adebug;
}

extern void MADDebugStr(short line, const char* file, const char* text)
{
	if (privateDebugStr) {
		(*privateDebugStr)(line, file, text);
	} else {
		fprintf(stderr, "%s:%u error text:%s!", file, line, text);
		fflush(stderr);
		
		abort();
	}
}

File: Classes/MyDebugStr.m
//
//  MyDebugStr.m
//  PPMacho
//
//  Created by C.W. Betts on 7/29/10.
//  Copyright 2010 __MyCompanyName__. All rights reserved.
//

#import <Cocoa/Cocoa.h>
#import <Carbon/Carbon.h>

extern __attribute__((visibility("default"))) void MyDebugStr( short line, Ptr file, Ptr text)
{
	NSLog(@"%s:%u error text:%s!", file, line, text);
	NSInteger alert = NSRunAlertPanel(NSLocalizedString(@"MyDebugStr_Error", @"Error"), 
	[NSString stringWithFormat:NSLocalizedString(@"MyDebugStr_MainText", @"The Main text to display"), text], 
	NSLocalizedString(@"MyDebugStr_Quit", @"Quit"), NSLocalizedString(@"MyDebugStr_Continue", @"Continue"), 
	NSLocalizedString(@"MyDebugStr_Debug", @"Debug"));
	switch (alert) {
		case NSAlertAlternateReturn:
			break;
		case NSAlertOtherReturn:
			Debugger();
			break;
		case NSAlertDefaultReturn:
			NSLog(@"Choosing to fail!");
		default:
			exit(EXIT_FAILURE);
			break;
	}
}
